Add PasswordDeriveBytes and CryptDeriveKey methods

steveharter opened the pull request:
Addresses issues:
https://github.com/dotnet/corefx/issues/12966 Port Rfc2898DeriveBytes.CryptDeriveKey
https://github.com/dotnet/corefx/issues/11118 Port T:System.Security.Cryptography.PasswordDeriveBytes

For Rfc2898DeriveBytes.CryptDeriveKey, PNSE exception is raised due to that type being in Algorithms and requirement that it calls CAPI (not CNG). The comment on that method explains it in more detail.

For PasswordDeriveBytes, two native methods needed to be ported (CapiHelper.DeriveKey and UnloadKey) from desktop.

@bartonjs please review.

steveharter commented on src/System.Security.Cryptography.Csp/src/System.Security.Cryptography.Csp.csproj:
```diff
@@ -11,24 +11,26 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <IsPartialFacadeAssembly Condition="'$(TargetGroup)' == 'net463'">true</IsPartialFacadeAssembly>
     <ResourcesSourceOutputDirectory Condition="'$(TargetGroup)' == 'net463'">None</ResourcesSourceOutputDirectory>
+    <NoWarn Condition="'$(TargetsUnix)' == 'true'">CS0809</NoWarn>
```
This is to prevent a build error for this auto-generated file. The source and ref files have the corresponding #pragma. Short of maintaining the unsupported.cs file by hand, this seems reasonable especially since it only applies to the Unix\PlatformNotSupported assembly.

justinvp commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs:
```diff
@@ -1415,6 +1416,163 @@ public static bool VerifySign(SafeProvHandle hProv, SafeKeyHandle hKey, int calg
+        // WARNING: This function side-effects its first argument (hProv)
+        private static void UnloadKey(SafeProvHandle hProv, SafeKeyHandle hKey, ref byte[] key_out, ref int cb_out)
+        {
+            byte[] rgbPubKey =
```
Any reason this couldn't be cached in a static readonly field?

steveharter replied:
Changed it to static readonly. The CryptImportKey documentation does not say it may modify.

bartonjs commented on src/System.Security.Cryptography.Csp/tests/PasswordDeriveBytesTests.cs:
```diff
@@ -0,0 +1,452 @@
+        }
+
+        [Fact]
+        public static void GetBytes_IdempotentIfReset()
```
I'd go with "StableIfReset", since Reset doesn't make the function become Idempotent

steveharter replied:
Done

bartonjs replied:
The name appears to still be IdempotentIfReset, though you claim to have changed it...

bartonjs commented on src/System.Security.Cryptography.Csp/tests/PasswordDeriveBytesTests.cs:
```diff
@@ -0,0 +1,452 @@
+            }
+
+            // Due to bug in GetBytes() these won't be equal. The bug is fixed in Rfc2898DeriveBytes.
+            Assert.NotEqual(first, second);
```
Is this the `The second parameter should really be _extraCount` bug?

steveharter replied:
Yes; updated comment to make that clear

bartonjs commented on src/System.Security.Cryptography.Csp/tests/PasswordDeriveBytesTests.cs:
```diff
@@ -0,0 +1,452 @@
+            byte[] key = TestKnownValue_CryptDeriveKey(
+                TestPassword,
+                "RC2",
+                "SHA1",
```
All the knownvalues tests seem to use SHA1.  Is it the only thing supported? If not, please add a test showing more than one value is valid.

bartonjs commented on src/System.Security.Cryptography.Csp/tests/PasswordDeriveBytesTests.cs:
```diff
@@ -0,0 +1,452 @@
+        {
+            byte[] output;
+
+            using (var deriveBytes = new PasswordDeriveBytes(password, salt, "SHA1", iterationCount))
```
Is SHA1 the only valid algorithm?  It won't take MD5 or SHA256?

steveharter replied:
Added MD5 and SHA256 to the RC2 tests

bartonjs replied:
I see MD5 and SHA256 were added to CryptDeriveKey.  This is for the PBKDF1-ish GetBytes... if we support other algorithms there we should validate that replacing the SHA1 identifier works.

bartonjs commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/PasswordDeriveBytes.cs:
```diff
@@ -0,0 +1,327 @@
+        private void HashPrefix(CryptoStream cs)
+        {
+            if (_prefix > 999)
+                throw new CryptographicException(SR.Cryptography_PasswordDerivedBytes_TooManyBytes);
```
I didn't see a test for this, can we make one?

steveharter replied:
Done

bartonjs commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/PasswordDeriveBytes.cs:
```diff
@@ -0,0 +1,327 @@
+                {
+                    // Note: The second parameter should really be _extraCount instead.
+                    // However, changing this would constitute a breaking change.
+                    Buffer.BlockCopy(_extra, ib, rgbOut, 0, ib);
```
The comment's a really good marker here, but if we don't have a test explicitly ensuring it doesn't change, can we add one?

steveharter replied:
There is an assert that tests this indirectly. You refer to it below; did not add another test.

bartonjs commented on src/System.Security.Cryptography.Csp/tests/PasswordDeriveBytesTests.cs:
```diff
@@ -276,6 +276,19 @@ public static void GetBytes_StreamLike()
+                deriveBytes.GetBytes(1000 * 20);
+
+                // Boundary case failure
+                Assert.Throws<CryptographicException>(() => deriveBytes.GetBytes(1001 * 20));
```
Doesn't it save the block count as a field, so only one needs to be called here?

steveharter replied:
Good catch (although test still passes); either I need a Reset() between them or just do 1 byte. Changes to do 1 byte.

steveharter commented:
@bartonjs any additional comments?

bartonjs commented on src/System.Security.Cryptography.Csp/tests/PasswordDeriveBytesTests.cs:
```diff
@@ -0,0 +1,481 @@
+                Buffer.BlockCopy(secondSecondHalf, 0, second, secondFirstHalf.Length, secondSecondHalf.Length);
+            }
+
+            // Due to "_extraCount" bug in GetBytes() these won't be equal. The bug is fixed in Rfc2898DeriveBytes.
```
If I understand that bug correctly, if the hash algorithm output size is requested then this does behave like a stream.  So with SHA1 if all requests are multiples of 20 bytes it should be like requesting everything in multiples of 40 bytes.  That edge case is probably good to capture, because (like all corner cases) someone might be depending on it.

steveharter replied:
I did add the happy path and compat tests for this to make this more clear. Yes if you request less than the hash size on the first call to GetBytes() then the "extra" bytes bug comes into play in the next call to GetBytes()

steveharter commented:
>> The name appears to still be IdempotentIfReset, though you claim to have changed it...

Renamed to StableIfReset (wrong method renamed)

>> I see MD5 and SHA256 were added to CryptDeriveKey. This is for the PBKDF1-ish GetBytes... if we support other algorithms there we should validate that replacing the SHA1 identifier works.

Added MD5 and SHA256 tests for GetBytes()

bartonjs commented on src/System.Security.Cryptography.Csp/tests/PasswordDeriveBytesTests.cs:
```diff
@@ -243,36 +243,77 @@ public static void GetBytes_IdempotentIfReset()
             }
 
-            // Due to "_extraCount" bug in GetBytes() these won't be equal. The bug is fixed in Rfc2898DeriveBytes.
+            // Since we requested >= 20 bytes in secondFirstHalf there no 'extra' bytes and we don't encounter the
```
Is ">=" accurate? or is it mod?

steveharter replied:
I did tweak the comment. It is >= 20 bytes for secondFirstHalf that prevents the extra bytes and the fact the that the second call for the remainder gets all of the remaining.

For example if we call GetBytes(22) followed by GetBytes(18) that would go into the `if (ib >= cb)` (ib = 18; cb = 18) block which doesn't have the bug.

So it's not necessarily a mod 20 that prevents the bug although mod 40 total would be close (however the second call could ask for more than 40 total but the extra bytes would be ignored as we only do two passes here)

steveharter commented:
The branch became corrupt; added new PR https://github.com/dotnet/corefx/pull/14112
